chanzuckerberg/corpora-data-portal
# thuang-909-validate-csv

Live link:

https://thuang-frontend.rdev.single-cell.czi.technology/?gs=true

#### Reviewers
**Functional:** 
@MDunitz 

**Readability:** 
@Bento007 

---

## Changes
- add
1. Dependency: `jschardet` and `papaparse` for checkig CSV encoding and parsing CSV rows respectively
2. Add `parseGeneSetsCSV()` util function to parse a CSV file. This is temporarily used on homepage behind feature flag `gs=true` for us to test out different CSV files and also for E2E tests
3. `parseGeneSetsCSV()` is responsible for both validating CSV input and checking the encoding of the CSV file. The output is parsed and converted result that matches BE's payload format and error messages
4. `validateOutput()` is used by `parseGeneSetsCSV()` to validate the CSV requirements shown in [this epic](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/115). Each validator is encapsulated as a function. E.g., `hasGeneSetName()`, `hasGeneSetDescription()`, `hasGeneSymbol()`. The helper function also injects other CSV parsing errors reported by library `papaparse`
5. `frontend/tests/features/geneSets.test.ts` runs E2E tests on different CSV files

Example result using `/fixtures/geneSets/Tidy CSV With Comments, empty gene set name, empty gene set description, empty gene symbol, empty gene description, duplicate gene set name, and duplicate gene symbol UTF8.csv`:

<img width="1529" alt="Screen Shot 2021-03-18 at 2 11 42 PM" src="https://user-images.githubusercontent.com/6309723/111698335-f27ccd00-87f3-11eb-832f-2c816d3fff49.png">

## Known Issues:
1. we need to **allow ASCII**, since If an `UTF-8` CSV file only has latin letters, it's the same as ASCII and `jschardet` will detect it as `ASCII` instead of `UTF-8`. [Source](https://www.azavea.com/blog/2014/03/24/solving-unicode-problems-in-python-2-7/#1_str_is_for_bytes_not_strings)

## Definition of Done (from ticket)
https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/909


- [x] Create test gene set files containing multiple gene sets representing the validation errors above
- [x] Test that all of the validation errors are caught and the error response specifies the gene set name and the validation error
- [x] All validation tests should be run each time (don't stop when one fails to ensure uploading gene sets isn't an endless nightmare of errors)

## QA steps (optional)

1. Pull branch
2. `npm i && npm start`
3. Visit: `http://localhost:8000/?gs=true`
4. Validate any CSV files in `frontend/tests/features/fixtures/geneSets/*` using the select file button
5. Verify the output is as expected
6. Repeat step 4 with other gene set CSV files you have and verify the output is as expected 🙏
